### PR TITLE
Hats!

### PR DESCRIPTION
**Pull request by AustinKellar**
Contributors: Austin, Wesley

Goals: 
- Introduce an expandable system for selecting hats and spawning them on players once in the match
- Model the new hat system after the skin system so that they are similar to modify / expand

What Changed:
- ControllerSignIn.cs
    - Removed the ability to change skins with Y
    - Added ability to cycle through skins in 2 directions with the Left and Right DPad buttons
    - Added in listeners for Up and Down DPad buttons to cycle through hats
    - Added in very simple keyboard controls so that Wesley can see what hats look like in Unity (H and J to cycle on keyboard)
- Hat.cs
    - A class to store Hat data and a function for putting a hat on a player
    - This class is very similar in spirit to Skin.cs
- PlayerHeadLocation.cs
    - This script literally just exists so that I can get the location of the top of Skelegoo's head through scripts in a clean way.
    - This script allows me to call `player.GetComponent<PlayerHeadLocation>()` to get that particular spot.
- PlayerSpawner.cs
    - Added in a call to SetStartingHat() when a player first spawns in
- ActivePlayers.cs
    - Added in a variable to store the hats that players chose in the lobby so that it is remembered throughout the game
-  PlayerColorController.cs --> PlayerCosmeticSetup.cs
    - I renamed PlayerColorController.cs to PlayerCosmeticSetup.cs 
    - This was the file that set the players' skins when they are first spawned into the match scene
    - I added functionality to this file so that it also spawns in the player hats, so I figured a rename would be appropriate
- UIController.cs
    - A race condition error appeared (once again), and I noticed that the Initialize() function was moved from Awake to Start, which was causing the SceneCreator to crash. I moved it back into Awake and renamed Initialize() to OnAwake() to avoid any further confusion.
- Temporary tutorial UI was added to the  lobby
- I added in a basic visual effect for changing hats. Subject to change in the future.
- Dummy Player Prefab:
    - I had to remake the Dummy Player prefab so that the win screen could display the hats
    - The only difference is the new prefab has the PlayerHeadLocation script on it.
- I had to alter the WinScreenController object for this PR, since I had to change out the Dummy Player prefab for the new one.
    - Hopefully this doesn't cause any nasty merge conflicts

How to Test:
- Hop into the lobby and make sure you can:
    - No longer change skins with Y
    - Cycle through skins with left and right dpad
    - Cycle through hats with up and down dpad 
        - Unlike skins, hats are not unique to one player. This means 2 players can have the same hat. This was a design deceision on my end, open to change if need be.
    - Spawn in and observe that the default hat is no hat.
    - Dab and ensure that the hat follows the skull
- Start a game and ensure that all players were spawned in with their chosen hats
- Win a game and ensure that all players were spawned in with their chosen hats (in win screen)
- Test all of the above in a build
- Start the match scene without going through the lobby and ensure that it is still playable.
    - Players should all spawn in with no hat.

NOTE: 
- This PR introduced a new problem where some of the hats block the text of the win screen.
- I feel that it is out of the scope of this feature, but we will need to at least slightly redesign the win screen to compensate for this.

### CODE REVIEW

File: Assets/Scripts/Lobby/State Management/PlayerSpawner.cs [hunk 1]
```diff
@@ -17,6 +17,7 @@ public class PlayerSpawner : MonoBehaviour
     private MaterialPropertyBlock _block;
```
Comment (IanGRap) [inline Assets/Scripts/Lobby/State Management/PlayerSpawner.cs]: we should require the PlayerHatChanger component

File: Assets/Scripts/Scene Construction/SceneCreator.cs [hunk 2]
```diff
@@ -14,11 +14,11 @@ public class SceneCreator : MonoBehaviour
     [SerializeField]
```
Comment (IanGRap) [inline Assets/Scripts/Scene Construction/SceneCreator.cs]: require PlayerCosmeticSetup?

File: Assets/Scripts/UI/Core/UIController.cs [hunk 3]
```diff
@@ -38,11 +38,15 @@ private void OnEnable()
     {
         OnMenuEnabled();
     }
-    
+
```
Comment (IanGRap) [inline Assets/Scripts/UI/Core/UIController.cs]: why is this file changed? I got some errors that might be related to these changes. 
hats error

File: Assets/Scripts/UI/Lobby/UIOptionsMenu.cs [hunk 4]
```diff
@@ -44,7 +44,7 @@ public class UIOptionsMenu : UIController
     private bool _changesMade = false;
     private enum MenuDirection { Left, Right };
 
-    protected override void Initialize()
+    protected override void OnAwake()
```
Comment (IanGRap) [inline Assets/Scripts/UI/Lobby/UIOptionsMenu.cs]: why is this file changed?

File: Assets/Scripts/UI/Lobby/UITitleMenu.cs [hunk 5]
```diff
@@ -21,7 +21,7 @@ public class UITitleMenu : UIController
     [SerializeField]
     private Button _quitButton;
 
-    protected override void Initialize()
+    protected override void OnAwake()
```
Comment (IanGRap) [inline Assets/Scripts/UI/Lobby/UITitleMenu.cs]: why is this file changed?

File: Assets/Scripts/UI/Match/UIPauseMenu.cs [hunk 6]
```diff
@@ -17,7 +17,7 @@ public class UIPauseMenu : UIController
 
     private UIMatchMenuInput _matchMenuInput;
 
-    protected override void OnMenuStart()
+    protected override void OnStart()
```
Comment (IanGRap) [inline Assets/Scripts/UI/Match/UIPauseMenu.cs]: why is this file changed?

File: Assets/Scripts/UI/Match/UIScoreMenu.cs [hunk 7]
```diff
@@ -8,7 +8,7 @@ public class UIScoreMenu : UIController
     [SerializeField]
     private Button _nextMatchButton;
 
-    protected override void Initialize()
+    protected override void OnAwake()
```
Comment (IanGRap) [inline Assets/Scripts/UI/Match/UIScoreMenu.cs]: why is this file changed?

File: Assets/Scripts/UI/Win Screen/UIWinScreenMenu.cs [hunk 8]
```diff
@@ -20,7 +20,7 @@ public class UIWinScreenMenu : UIController
 
     private bool _isExecutingAction = false;
 
-    protected override void OnMenuStart()
+    protected override void OnStart()
```
Comment (IanGRap) [inline Assets/Scripts/UI/Win Screen/UIWinScreenMenu.cs]: why is this file changed?

File: Assets/Scripts/UI/Core/UIController.cs (hunk 3 above)
Comment (AustinKellar) [reply Assets/Scripts/UI/Core/UIController.cs]: After not touching the UI, I was getting a race condition where the SceneCreator was relying on something on Start of the UI, and Alfredo moved the UI function from Awake (where it's supposed to be) to Start. This introduces a race condition when Unity decides on a certain order of running Start functions.

File: Assets/Scripts/UI/Lobby/UIOptionsMenu.cs (hunk 4 above)
Comment (AustinKellar) [reply Assets/Scripts/UI/Lobby/UIOptionsMenu.cs]: I layed it out in the PR comments. I ran into a race condition involving the Initialize function being on Start instead of Awake (where it is intended to be) so I moved it where it should be and renamed it so it is clear.

File: Assets/Scripts/UI/Lobby/UITitleMenu.cs (hunk 5 above)
Comment (AustinKellar) [reply Assets/Scripts/UI/Lobby/UITitleMenu.cs]: Same as the others. The rename of the function.

File: Assets/Scripts/UI/Core/UIController.cs (hunk 3 above)
Comment (AustinKellar) [reply Assets/Scripts/UI/Core/UIController.cs]: I'll look into this new error. I didn't encounter it myself.

Comment (AustinKellar) [reply Assets/Scripts/UI/Core/UIController.cs]: The error in question is the same race condition that has popped up many times throughout the project. When Unity decides to change up the order it runs Start functions, this occasionally pops up. That's why I had to change the UI code.

File: Assets/Scripts/UI/Match/UIPauseMenu.cs (hunk 6 above)
Comment (AustinKellar) [reply Assets/Scripts/UI/Match/UIPauseMenu.cs]: You should read the PR description. I explained why I thought this was necessary.

File: Assets/Scripts/UI/Core/UIController.cs (hunk 3 above)
Comment (IanGRap) [reply Assets/Scripts/UI/Core/UIController.cs]: so it looks like by fixing one race condition we created another

File: Assets/Scripts/UI/Match/UIPauseMenu.cs (hunk 6 above)
Comment (IanGRap) [reply Assets/Scripts/UI/Match/UIPauseMenu.cs]: I had run into a problem on my own branches where certain files weren't updating and I had mistakenly assumed that was happening here. I just meant the comment as flagging files that I thought hadn't been updated with dev. I apologize for the obnoxious comments.

File: Assets/Scripts/UI/Core/UIController.cs (hunk 3 above)
Comment (AustinKellar) [reply Assets/Scripts/UI/Core/UIController.cs]: Yeah, can you give a more detailed step-by-step? I haven't been able to reproduce the one you found.

Comment (IanGRap) [reply Assets/Scripts/UI/Core/UIController.cs]: so I started in the lobby, then played through a match. On the win screen, I chose to go back to the lobby. Once I was on the lobby, I wasn't able to choose menu options and got that error.

File: Assets/Scripts/UI/Core/UIController.cs [hunk 9]
```diff
@@ -60,6 +60,14 @@ private void Update()
     /// <para>Called in Unity's Start() event function</para>
     /// </summary>
     protected virtual void OnAwake()
+    {
```
Comment (IanGRap) [inline Assets/Scripts/UI/Core/UIController.cs]: remove OnAwake()?

Comment (AustinKellar) [reply Assets/Scripts/UI/Core/UIController.cs]: Its a virtual function and the other classes that inherit from this need access to it.

File: Assets/Scripts/UI/Core/UIController.cs [hunk 10]
```diff
@@ -55,7 +59,15 @@ private void Update()
     /// Should be used to initialize any values needed by the UI object
     /// <para>Called in Unity's Start() event function</para>
```
Comment (AlfredoCuevas) [inline Assets/Scripts/UI/Core/UIController.cs]: This comment needs to be updated

File: Assets/Scripts/UI/Core/UIController.cs [hunk 11]
```diff
@@ -55,7 +59,15 @@ private void Update()
...
     /// </summary>
-    protected virtual void Initialize()
+    protected virtual void OnAwake()
+    {
+        
+    }
+
+    /// <summary>
+    /// Called after Initialize() inside Unity's Start() event function
```
Comment (AlfredoCuevas) [inline Assets/Scripts/UI/Core/UIController.cs]: This comment also needs to be updated.
